File: 0874-backspace-string-compare/0874-backspace-string-compare.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
        int i = s.length() - 1;
        int j = t.length() - 1;
        while(i >= 0 || j >= 0){
            i = nextValidIndex(s, i);
            j = nextValidIndex(t, j);

            if(i < 0 && j < 0) return true;
            if(i < 0 || j < 0 || s.charAt(i) != t.charAt(j)) return false;

            i--;
            j--;
        }
        return true;
    }
    
    private int nextValidIndex(String str, int index){
        int backCount = 0;
        while(index >= 0){
            if(str.charAt(index) == '#'){
                backCount++;
                index--;
            }else if(backCount > 0){
                backCount--;
                index--;
            }else{
                break;
            }
        }
        return index;
    }
}

File: 0874-backspace-string-compare/0874-backspace-string-compare.py
class Solution(object):
    def backspaceCompare(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        p1 = len(s) - 1
        p2 = len(t) - 1

        skipS = 0
        skipT = 0
        while p1 >=0 or p2 >= 0:
            while p1 >= 0:
                if s[p1] == '#':
                    skipS += 1
                    p1 -= 1
                elif skipS > 0:
                    skipS -= 1
                    p1 -= 1
                else:
                    break
            while p2 >= 0:
                if t[p2] == '#':
                    skipT += 1
                    p2 -= 1
                elif skipT > 0:
                    skipT -= 1
                    p2 -= 1
                else:
                    break
            if p1 >= 0 and p2 >= 0 and s[p1] != t[p2]:
                return False
            if (p1 >= 0) != (p2 >= 0):
                return False
            p1 -= 1
            p2 -= 1
        return True